### PR TITLE
Add parameters to queries

### PR DESCRIPTION
**Pull request by 4ice**
* Add all documented parameters to the match operator.
* Adds minimum_should_match to bool query - This is required in order to combine "should" and "filter".

### CODE REVIEW

**Comment by Jeroen-G** [comment]
Awesome! I'll find time to take a closer look and I have approved the workflows to run - I think you will need to add some test cases though.

**Comment by 4ice** [comment]
Great that you liked it!

I am a beginner at testing, but I have modified the tests that was required to change, and I have also added new tests for the new functions for matching. If there is something missing, or that is wrong, tell me and I will try to fix it! :)

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,5 @@
 .phpunit.result.cache
 composer.lock
 infection.log
+
+.idea
```
Comment (Jeroen-G) [inline .gitignore:8]: This should be in your global .gitignore.

**Comment by Jeroen-G** [comment]
> I am a beginner at testing

Then you are up for a treat because I gave myself the challenge with this package that it must have 100% coverage + 100% mutation coverage 😄

**Comment by Jeroen-G** [comment]
You can run phpunit and infection locally with the make commands if you don't want to wait for the github actions.

File: .gitignore (hunk 1 above)
Comment (4ice) [reply .gitignore]: Okay, fixed!

**Comment by 4ice** [comment]
Aha, I just ran phpunit, I will try out infection and figure out those parts as well!

**Comment by 4ice** [comment]
Also, there are two cases of "Not Covered mutants" that is not part of my commit, I think at least...

```
Not Covered mutants:
====================

1) /Explorer/src/Domain/Analysis/Analyzer/StandardAnalyzer.php:35    [M] InstanceOf_

--- Original
+++ New
@@ @@
     public function getFilters() : array
     {
         return array_map(function ($filter) {
-            if ($filter instanceof FilterInterface) {
+            if (true) {
                 return $filter->getName();
             }
             return $filter;


2) /Explorer/src/Domain/Analysis/Analyzer/StandardAnalyzer.php:35    [M] InstanceOf_

--- Original
+++ New
@@ @@
     public function getFilters() : array
     {
         return array_map(function ($filter) {
-            if ($filter instanceof FilterInterface) {
+            if (false) {
                 return $filter->getName();
             }
             return $filter;
```

**Comment by Jeroen-G** [comment]
> I have gone through the test cases with infection as well. I have managed to fix everything except one case:
> 
> ```
> Escaped mutants:
> ================
> 
> 1) /Explorer/src/Infrastructure/Scout/ScoutSearchCommandBuilder.php:68    [M] MethodCallRemoval
> 
> --- Original
> +++ New
> @@ @@
>          $normalizedBuilder->setFields($builder->fields ?? []);
>          $normalizedBuilder->setBoolQuery($builder->compound ?? new BoolQuery());
>          $normalizedBuilder->setLimit($builder->limit);
> -        $normalizedBuilder->setMinimumShouldMatch($builder->minimumShouldMatch ?? null);
> +        
>          $index = $builder->index ?: $builder->model->searchableAs();
>          if ($builder->model instanceof SearchableFields) {
>              $normalizedBuilder->setDefaultSearchFields($builder->model->getSearchableFields());
> ```
> 
> I can't figure out how to solve this, could you give me a hand? :)

I see that the mutant is removing that line. No test fails, so no test is testing that line. I think you should write a test that uses `setMinimumShouldMatch` to set it to a custom value.

**Comment by Jeroen-G** [comment]
> Also, there are two cases of "Not Covered mutants" that is not part of my commit, I think at least...

That is strange! But you seem to be right. So if you don't want to test them then that's fine. I can do it later.

**Comment by 4ice** [comment]
> > I have gone through the test cases with infection as well. I have managed to fix everything except one case:
> > ```
> > Escaped mutants:
> > ================
> > 
> > 1) /Explorer/src/Infrastructure/Scout/ScoutSearchCommandBuilder.php:68    [M] MethodCallRemoval
> > 
> > --- Original
> > +++ New
> > @@ @@
> >          $normalizedBuilder->setFields($builder->fields ?? []);
> >          $normalizedBuilder->setBoolQuery($builder->compound ?? new BoolQuery());
> >          $normalizedBuilder->setLimit($builder->limit);
> > -        $normalizedBuilder->setMinimumShouldMatch($builder->minimumShouldMatch ?? null);
> > +        
> >          $index = $builder->index ?: $builder->model->searchableAs();
> >          if ($builder->model instanceof SearchableFields) {
> >              $normalizedBuilder->setDefaultSearchFields($builder->model->getSearchableFields());
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I can't figure out how to solve this, could you give me a hand? :)
> 
> I see that the mutant is removing that line. No test fails, so no test is testing that line. I think you should write a test that uses `setMinimumShouldMatch` to set it to a custom value.

Yes, that was my though as well, but I have the following test: https://github.com/Jeroen-G/Explorer/pull/68/files#diff-81e8f81bdd3cc55a3be531518fcf54ab4b58434b6f0e474aa5b987a254c6c020R232 which uses `setMinimumShouldMatch` and sets it to "50%", that test passes.

**Comment by Jeroen-G** [comment]
If you merge back or rebase from master the actions should work again (or at least really run and check the code).

**Comment by 4ice** [comment]
@Jeroen-G - I have rebased against master. Could you approve running the workflow?

**Comment by 4ice** [comment]
Woho, all checks have passed! Thanks for making dip my toes into testing :D Do you need anything from me before merging?

**Comment by Jeroen-G** [comment]
There are a few escaped mutants according to the review bot (see code tab), could you take a look at those again?

**Comment by Jeroen-G** [comment]
And thank YOU for your contributions so far!
I should have time to test and merge this before the weekend. Then it'll be in main to allow you to test it and lastly I'll tag the release.

**Comment by 4ice** [comment]
> There are a few escaped mutants according to the review bot (see code tab), could you take a look at those again?

Yeah, the escaped mutant mention there is the one we discussed earlier in this PR. I can't seem to figure out how to get rid of that. I have the following test: https://github.com/Jeroen-G/Explorer/pull/68/files#diff-81e8f81bdd3cc55a3be531518fcf54ab4b58434b6f0e474aa5b987a254c6c020R232 which uses setMinimumShouldMatch and sets it to "50%", that test passes. Do you have any ideas why that does not solve the escaped mutant?

**Comment by Jeroen-G** [comment]
I have one last functional question: instead of `'minimum_should_match' => null,` should we not just leave it out?
Is setting it to null affecting anything?

**Comment by 4ice** [comment]
I am not sure I follow.. The places where I have added `'minimum_should_match' => null` is in "expects"-statements of tests, because i expect that minimum_should_match to be null if not set. If I try to remove it from the "expects" the phpunit test fails. Or am I missing something? :)

**Comment by Jeroen-G** [comment]
What is the difference between a query with `'minimum_should_match' => null` and a query with no `minimum_should_match` at all?

**Comment by 4ice** [comment]
Hi again! Thank you for your feedback, I have resolved your comment, and I also managed to fix the infection mutatant that was an issue before. All metrics are 100% now! Could you please review it again an run the workflow. :)

**Pull request merged**